File: python/bintree.py
class BinTree(object):
    """
    Non self balancing Binary Tree
    """
    def __init__(self, value=None):
        self.left_child = None
        self.right_child = None
        self.value = value

    def __len__(self):
        return self.length()

    def has_children(self):
        return self.left_child or self.right_child

    def insert_many(self, *values):
        for v in values:
            self.insert(v)

    def insert(self, value):
        if not self.value:
            self.value = value
            return self
        elif value > self.value:
            if self.right_child is None:
                self.right_child = BinTree(value)
                return self.right_child
            else:
                self.right_child.insert(value)
        elif value < self.value:
            if self.left_child is None:
                self.left_child = BinTree(value)
                return self.left_child
            else:
                self.left_child.insert(value)
        return self  # if self.value == value

    def search(self, value):
        if value == self.value:
            return self
        elif value > self.value and self.right_child:
            return self.right_child.search(value)
        elif value < self.value and self.left_child:
            return self.left_child.search(value)
        return -1

    def remove(self, value):
        if value == self.value:
            if self.has_children():  # this guarantees at least one child
                rc = self.right_child.depth() if self.right_child else 0
                lc = self.left_child.depth() if self.left_child else 0
                if rc > lc or self.left_child is None:
                    self.value = self.right_child.minimum().value
                    self.right_child.remove(self.value)
                else:
                    self.value = self.left_child.maximum().value
                    self.left_child.remove(self.value)
            else:
                # Set it to None and purge it in higher recurrency level.
                self.value = None
        elif value > self.value:
            if self.right_child:
                self.right_child.remove(value)
        elif value < self.value:
            if self.left_child:
                self.left_child.remove(value)
        # Purging Nones
        if self.left_child is not None and self.left_child.value is None:
            self.left_child = None
        if self.right_child is not None and self.right_child.value is None:
            self.right_child = None
        return self

    def length(self):
        if not self.value:
            return 0
        res = 1
        if self.left_child:
            res += self.left_child.length()
        if self.right_child:
            res += self.right_child.length()
        return res

    def depth(self):
        if not self.value:
            return 0
        lvl, lc, rc = 1, 0, 0
        if self.left_child:
            lc += self.left_child.depth()
        if self.right_child:
            rc += self.right_child.depth()
        return lvl + max(lc, rc)

    def minimum(self):
        if self.left_child is None:
            return self
        return self.left_child.minimum()

    def maximum(self):
        if self.right_child is None:
            return self
        return self.right_child.maximum()

    def to_dict(self):
        repr = {'value': self.value}
        if self.right_child:
            repr['right_child'] = self.right_child.to_dict()
        if self.left_child:
            repr['left_child'] = self.left_child.to_dict()
        return repr

    def to_list(self):
        lc = self.left_child.to_list() if self.left_child else []
        rc = self.right_child.to_list() if self.right_child else []
        return lc + [self.value] + rc

    @staticmethod
    def from_list(array):
        alen = len(array)
        middle = alen//2
        lc = array[:middle]
        val = array[middle]
        rc = array[middle+1:]
        root = BinTree(val)
        if lc:
            root.left_child = root.from_list(lc)
        if rc:
            root.right_child = root.from_list(rc)
        return root

File: c/bintree.c
#include <stdio.h>
#include <stdlib.h>
#include "types.h"

binTree *getTree () {
    binTree *a = malloc(sizeof(binTree));;
    a->value = 0;
    a->rightChild = NULL;
    a->leftChild = NULL;
    return a;
}

void destroyTree(binTree *root) {
    if (root == NULL) {
        return;
    }
    destroyTree(root->leftChild);
    destroyTree(root->rightChild);
    printf("Destroying %d\n", root->value);
    free(root);
    return;
}

void insert(binTree *root, int value) {
    if (root->value == 0) {
        printf("New child populated with %d\n", value);
        root->value = value;
        return;
    }
    if (root->value < value) {
        if (root->rightChild == NULL) {
            printf("Right child created\n");
            root->rightChild = getTree();
        }
        printf("Populating right child with: %d\n", value);
        insert(root->rightChild, value);
    } else {
        if (root->leftChild == NULL) {
            printf("Left child created\n");
            root->leftChild = getTree();
        }
        printf("Populating left child with: %d\n", value);
        insert(root->leftChild, value);
    }
    return;
}

void insertMany(binTree *root, int argsc, int *args) {
    for (int i=0; i < argsc; i++) {
        insert(root, args[i]);
    }
}

int search(binTree *root, int value) {
    if (root == NULL) {
        return 0;
    }
    if (root->value == value) {
        printf ("Value %d found!\n", value);
        return 1;
    } else if (root->value > value) {
        return search(root->leftChild, value);
    } else if (root->value < value) {
        return search(root->rightChild, value);
    } else {
        return 0;
    }
}

int length(binTree *root) {
    if (root == NULL) {
        return 0;
    }
    int len = 1;
    len += length(root->leftChild);
    len += length(root->rightChild);
    return len;
}

int depth(binTree *root) {
    if (root == NULL) {
        return 0;
    }
    int dep = 1;
    int lc = depth(root->leftChild);
    int rc = depth(root->rightChild);
    dep += lc > rc ? lc : rc;
    return dep;
}

int minimum(binTree *root) {
    if (root == NULL) {
        return -1;
    }
    if (root->leftChild != NULL) {
        return minimum(root->leftChild);
    } else {
        return root->value;
    }
}

int maximum(binTree *root) {
    if (root == NULL) {
        return -1;
    }
    if (root->rightChild != NULL) {
        return maximum(root->rightChild);
    } else {
        return root->value;
    }
}

void removeFromTree(binTree *root, int value) {
    if (root == NULL) {
        return;
    }
    if (root->value == value) {
        if (root->leftChild != NULL || root->rightChild != NULL) {
            int ldep = depth(root->leftChild);
            int rdep = depth(root->rightChild);
            if (rdep > ldep || root->leftChild == NULL) {
                int val = minimum(root->rightChild);
                root->value = val;
                removeFromTree(root->rightChild, val);
            } else {
                int val = maximum(root->leftChild);
                root->value = val;
                removeFromTree(root->leftChild, val);
            }
        } else {
            root->value = 0;
        }
    } else if (root->value > value) {
        removeFromTree(root->leftChild, value);
    } else if (root->value < value) {
        removeFromTree(root->rightChild, value);
    }
    if (root->leftChild != NULL && root->leftChild->value == 0) {
        free(root->leftChild);
        root->leftChild = NULL;
    }
    if (root->rightChild != NULL && root->rightChild->value == 0) {
        free(root->rightChild);
        root->rightChild = NULL;
    }
    return;
}

binTree *fromList(int len, int *args) {
    int middle = (int)len/2;
    int value = *(args+middle);
    binTree *root = getTree();
    insert(root, value);
    int leftLen = middle;
    int rightLen = len - middle - 1;
    if (leftLen > 0) {
        root->leftChild = fromList(leftLen, args);
    }
    if (rightLen > 0) {
        root->rightChild = fromList(rightLen, args+middle+1);
    }
    return root;
}

linkedList *toList(binTree *root) {
    if (root == NULL) {
        return NULL;
    }
    linkedList *total = getLinkedList();
    if (root->leftChild != NULL) {
        linkedList *leftChild = toList(root->leftChild);
        joinLists(total, leftChild);
        free(leftChild);
    }
    addValueToList(total, root->value);
    if (root->rightChild != NULL) {
        linkedList *rightChild = toList(root->rightChild);
        joinLists(total, rightChild);
        free(rightChild);
    }
    printList(total);
    return total;
}
